File: substring.go
package main

import "fmt"

//A function which returns the length of longest substring.
//Which takes a String as input.
func SubString(inputString string) int {
	if len(inputString) == 0 {
		return 0
	}
	charLastIndex := map[rune]int{}
	currentSubStringLength, longestSubStringLength, start := 0, 0, 0
	for index, character := range inputString {
		if lastIndex, hasCharacter := charLastIndex[character]; !hasCharacter || lastIndex < index-currentSubStringLength {
			currentSubStringLength++
		} else {
			if currentSubStringLength > longestSubStringLength {
				longestSubStringLength = currentSubStringLength
			}
			start = lastIndex + 1
			currentSubStringLength = index - start + 1
		}
		charLastIndex[character] = index
	}
	if currentSubStringLength > longestSubStringLength {
		longestSubStringLength = currentSubStringLength
	}
	return longestSubStringLength
}

func main() {
	fmt.Println(SubString("abcabcbb"))
	fmt.Println(SubString("bbbbb"))
	fmt.Println(SubString("pwwkew"))
	fmt.Println(SubString("javaconceptoftheday"))
}

File: substring.ts
/**
 *
 * A function which returns the length of longest substring.
 * Which takes a String as input.
 *
 * @param  {string} inputString - Input String
 * @returns Length of the Longest Substring
 */
function lengthOfLongestSubstring(inputString: string): number {
  if (inputString.length === 0) {
    return 0;
  }
  let subString = new Set<string>();
  let start = 0,
    end = 0,
    maxLength = 0;
  while (start < inputString.length && end < inputString.length) {
    if (!subString.has(inputString[end])) {
      subString.add(inputString[end]);
      end++;
      maxLength = Math.max(maxLength, end - start);
    } else {
      subString.delete(inputString[start]);
      start++;
    }
  }
  return maxLength;
}

console.log(lengthOfLongestSubstring("abcabcbb"));
console.log(lengthOfLongestSubstring("bbbbb"));
console.log(lengthOfLongestSubstring("pwwkew"));
console.log(lengthOfLongestSubstring("javaconceptoftheday"));
